plugins/sharding/logical_enumerator.rb

```diff
@@ -22,6 +23,14 @@ def reverse_each(&block)
         each_internal(:descending, &block)
       end
 
+      def each_with_index(&block)
+        each_with_index_internal(:ascending, &block)
+      end
+
+      def reverse_each_with_index(&block)
+        each_with_index_internal(:descending, &block)
+      end
```

Comment:
You don't need them with the change:

```diff
diff --git a/plugins/sharding/logical_enumerator.rb b/plugins/sharding/logical_enumerator.rb
index 79e5f7003..5dde0fb8f 100644
--- a/plugins/sharding/logical_enumerator.rb
+++ b/plugins/sharding/logical_enumerator.rb
@@ -28,6 +28,7 @@ module Groonga
 
       private
       def each_internal(order)
+        return enum_for(__method__, order) unless block_given?
         context = Context.instance
         each_shard_with_around(order) do |prev_shard, current_shard, next_shard|
           shard_range_data = current_shard.range_data
```

Then you can use like `enumerator.each.with_index`.

Reply:
Great, I didn't know that!